src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -24,6 +24,45 @@ License
 \*---------------------------------------------------------------------------*/
 
 #include "directAMI.H"
+#include "primitiveFieldsFwd.H"
+#include "pointFieldFwd.H"
+#include "faceData.H"
+
+namespace Foam {
+    struct ShootRayFunctor : public std::binary_function<label, label, void> {
+        const pointField *tgtPoints;
+        const pointField *srcPoints;
+        const vectorField *srcCf;
+        const faceData *tgtFaces;
+        const faceData *srcFaces;
+
+        ShootRayFunctor
+        (
+        const pointField *_tgtPoints,
+        const pointField *_srcPoints,
+        const vectorField *_srcCf,
+        const faceData *_tgtFaces,
+        const faceData *_srcFaces
+        ) :
+        tgtPoints(_tgtPoints),
+        srcPoints(_srcPoints),
+        srcCf(_srcCf),
+        tgtFaces(_tgtFaces),
+        srcFaces(_srcFaces)
+        {}
+
+        __HOST____DEVICE__
+        void operator()(const label& tgtI, const label& srcI) {
+            pointHit ray = tgtFaces[tgtI].ray(srcCf[srcI], srcFaces[srcI].normal(srcPoints), tgtPoints);
+            if (ray.hit()) {
+                // found an intersection !
+                // Doesn't work currently since the "ray" method is
+                // commented out in faceData.H
+                // We should update the addressing here.
```

Comment:
The ray method was not yet ported to the GPU. It would be best to move this method to the faceFunctors.H header located in src/OpenFOAM/meshes/meshShapes/face. I would like to eventually get rid of the faceData class in favour of functors that have all the methods from faceData class implemented. This will help to reduce memory used for storing face information.